parlai/agents/language_model/language_model.py

```diff
@@ -0,0 +1,480 @@
+                    vecs_to_return += observe
+                dict_to_return = {'text': '', 'labels': '', 'text2vec': vecs_to_return}
+                self.observation = dict_to_return
+                return dict_to_return
+        else:
+            if 'text' in obs:
+                obs['text'] += ' __END__'
+            self.observation = obs
+            return obs
+
+    def repackage_hidden(self, h):
+        """Wraps hidden states in new Variables, to detach them from their history."""
+        if type(h) == Variable:
+            return Variable(h.data)
+        else:
+            return tuple(self.repackage_hidden(v) for v in h)
+
+    def get_target_loss(self, data, hidden, targets, y_lens):
+        """Calculates the loss with respect to the targets, token by token,
+           where each output token is conditioned on either the input or the
+           previous target token.
+        """
+        loss = 0.0
+        bsz = data.size(0)
+        # use y_lens to find out number of tokens in each loss calculation
+        y_lens_batch = [len(y_lens)]*targets.size(1)
+        for i in range(len(y_lens_batch)):
+            num_less = 0
+            for y_len in y_lens:
+                if (y_len) - 1 < i:
+                    num_less +=1
+            y_lens_batch[i] -= num_less
+
+        output, hidden = self.model(data.transpose(0,1), hidden)
+        self.hidden = self.repackage_hidden(hidden)
+        # get only last predicted word
+        output = output.select(0,-1)
+        output_flat = output.view(-1, len(self.dict))
+        loss += y_lens_batch[0]*self.criterion(output_flat, targets.select(1,0).view(-1)).data
+
+        for i in range(1, targets.size(1)):
+            output, hidden = self.model(targets.select(1,i-1).view(1, bsz), self.hidden, no_pack=True)
+            self.hidden = self.repackage_hidden(hidden)
+            output_flat = output.view(-1, len(self.dict))
+            loss += y_lens_batch[i]*self.criterion(output_flat, targets.select(1,i).view(-1)).data
+
+        return loss/float(sum(y_lens_batch))
+
+    def get_predictions(self, data, end_idxs):
+        """Generates predictions word by word until we either reach the end token
+           or some max length (opt['truncate_pred']).
+        """
+        token_list = []
+        bsz = data.size(0)
+        done = [False for _ in range(bsz)]
+        total_done = 0
+        hidden = self.model.init_hidden(bsz)
+        output, hidden = self.model(data.transpose(0,1), hidden)
+        hidden = self.repackage_hidden(hidden)
+        word_weights = output.squeeze().data.exp()
+        # get last word of output
+        if bsz > 1:
+            value, word_idx = torch.max(word_weights, 2)
+            last_words = []
+            for j in range(len(end_idxs)):
+                last_words.append(word_idx[end_idxs[j]-1][j])
+            word_idx = torch.LongTensor(last_words)
+            if self.use_cuda:
+                word_idx = word_idx.cuda()
```

Comment:
it'd be better if we could (1) never call `cuda()` and (2) never construct a Tensor in this function. we shouldn't need to--can't we just pass word_idx directly?

it looks like `word_idx[end_idxs[j]-1][j]` might be accessing a specific element not a sub-tensor, in which case it would copy that back off the GPU to the CPU. it'd be better if we could wait to copy things back from the GPU to the CPU until we finish completely--can we just append tensors to the list, concatenate them all at the end, and then copy that back? that should run faster as well.